patches/server/1005-PreSpawnerSearchEvent.patch

```diff
@@ -0,0 +1,61 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: TreemanKing <67459602+TreemanKing@users.noreply.github.com>
+Date: Wed, 16 Aug 2023 11:53:44 +1000
+Subject: [PATCH] PreSpawnerSearchEvent
+
+
+diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+index 633500aefd515df5dadda3802b94079f75a03fa0..6d22659cf13b6bef32f0d058a6489cf1d365d642 100644
+--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
++++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+@@ -1,6 +1,8 @@
+ package net.minecraft.world.level;
+ 
+ import com.mojang.logging.LogUtils;
++
++import java.util.Collection;
+ import java.util.Optional;
+ import java.util.function.Function;
+ import javax.annotation.Nullable;
+@@ -22,6 +24,7 @@ import net.minecraft.world.entity.Mob;
+ import net.minecraft.world.entity.MobSpawnType;
+ import net.minecraft.world.entity.SpawnGroupData;
+ import net.minecraft.world.entity.SpawnPlacements;
++import net.minecraft.world.entity.EntitySelector;
+ import net.minecraft.world.level.gameevent.GameEvent;
+ import net.minecraft.world.phys.AABB;
+ import org.slf4j.Logger;
+@@ -46,6 +49,7 @@ public abstract class BaseSpawner {
+     public int requiredPlayerRange = 16;
+     public int spawnRange = 4;
+     private int tickDelay = 0; // Paper
++    private int searchCooldownTicks = 0; // Paper
+ 
+     public BaseSpawner() {}
+ 
+@@ -86,7 +90,24 @@ public abstract class BaseSpawner {
+         tickDelay = world.paperConfig().tickRates.mobSpawner;
+         if (tickDelay == -1) { return; } // If disabled
+         // Paper end
+-        if (this.isNearPlayer(world, pos)) {
++
++        // Paper start - PreSpawnerSearchEvent
++        if (--searchCooldownTicks > 0) return;
++
++        Collection<org.bukkit.entity.Player> activatingPlayers = (Collection)world.getNearbyPlayers(null,
++            (double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double)
++                this.requiredPlayerRange, EntitySelector.PLAYER_AFFECTS_SPAWNING);
++
++        io.papermc.paper.event.entity.PreSpawnerSearchEvent preSpawnerSearchEvent;
++        preSpawnerSearchEvent = new io.papermc.paper.event.entity.PreSpawnerSearchEvent(
++            io.papermc.paper.util.MCUtil.toLocation(world, pos), activatingPlayers, searchCooldownTicks);
```

Comment:
This can be on the same line but the event doesn't seems to be called anywhere in the first place?

Reply:
What would you recommend alternatively? Isn't it required so that "setCancelled" can be checked?

Reply:
The event doesn't seems to be called anywhere normally you do a #callEvent or Bukkit...callEvent

Reply:
So something like this?
`if(!preSpawnerSearchEvent.callEvent()){
if(event.setCancelled()){ return } }`

Reply:
you dont need the second line iirc.